asserts/user.go

```diff
@@ -0,0 +1,212 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+	"net/mail"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var validSystemUserUsernames = regexp.MustCompile(`^[a-z0-9][-a-z0-9+.-_]*$`)
+
+// SystemUser holds a system-user assertion which allows creating local
+// system users.
+type SystemUser struct {
+	assertionBase
+	series  []string
+	models  []string
+	sshKeys []string
+	since   time.Time
+	until   time.Time
+}
+
+// BrandID returns the brand identifier that signed this assertion.
+func (su *SystemUser) BrandID() string {
+	return su.HeaderString("brand-id")
+}
+
+// EMail returns the email address that this assertion is valid for.
+func (su *SystemUser) EMail() string {
+	return su.HeaderString("email")
+}
+
+// Series returns the series that this assertion is valid for.
+func (su *SystemUser) Series() []string {
+	return su.series
+}
+
+// Models returns the models that this assertion is valid for.
+func (su *SystemUser) Models() []string {
+	return su.models
+}
+
+// Name returns the full name of the user (e.g. Random Guy).
+func (su *SystemUser) Name() string {
+	return su.HeaderString("name")
+}
+
+// Username returns the system user name that should be created (e.g. "foo").
+func (su *SystemUser) Username() string {
+	return su.HeaderString("username")
+}
+
+// Password returns the crypt(3) compatible password for the user.
+// Note that only ID: $6$ or stronger is supported (sha512crypt).
+func (su *SystemUser) Password() string {
+	return su.HeaderString("password")
+}
+
+// SSHKeys returns the ssh keys for the user.
+func (su *SystemUser) SSHKeys() []string {
+	return su.sshKeys
+}
+
+// Since returns the time since the assertion is valid.
+func (su *SystemUser) Since() time.Time {
+	return su.since
+}
+
+// Until returns the time until the assertion is valid.
+func (su *SystemUser) Until() time.Time {
+	return su.until
+}
+
+// ValidAt returns whether the system-user is valid at 'when' time.
+func (su *SystemUser) ValidAt(when time.Time) bool {
+	valid := when.After(su.since) || when.Equal(su.since)
+	if valid && !su.until.IsZero() {
+		valid = when.Before(su.until)
+	}
+	return valid
+}
+
+// Implement further consistency checks.
+func (su *SystemUser) checkConsistency(db RODatabase, acck *AccountKey) error {
+	// Do the cross-checks when this assertion is actually used,
+	// i.e. in the create-user code. See also Model.checkConsitency
+
+	return nil
+}
+
+// sanity
+var _ consistencyChecker = (*SystemUser)(nil)
+
+func checkHashedPassword(headers map[string]interface{}, name string) (string, error) {
+	pw, err := checkOptionalString(headers, name)
```

Comment:
why optional?  we check it

do we want a way to support "*" (I think) to express not being able to login with a password? but just ssh keys